### PR TITLE
[SPARK-12567][SQL] Add aes_encrypt and aes_decrypt builtin functions

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR adds AES encryption/decryption builtin functions for SQL called `aes_encrypt` and `aes_decrypt`.
MySQL, Oracle Database and Hive implement this feature.

### Why are the changes needed?

Like hash functions, encryption/decryption is common for data processing.
Hive also implements these functions.

### Does this PR introduce _any_ user-facing change?

Yes. New functions will be available in SQL.

### How was this patch tested?

New tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 1]
```diff
@@ -301,3 +305,137 @@ case class CurrentUser() extends LeafExpression with Unevaluable {
...
+/**
+ * A function that encrypts input using AES. Key lengths of 128, 192 or 256 bits can be used.
+ * For versions prior to JDK 8u161, 192 and 256 bits keys can be used
+ * if Java Cryptography Extension (JCE) Unlimited Strength Jurisdiction Policy Files are installed.
+ * If either argument is NULL or the key length is not one of the permitted values,
+ * the return value is NULL.
+ */
+@ExpressionDescription(
+  usage = """
+    _FUNC_(expr, key) - Returns a encrypted value of `expr` using AES.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala:398]: nit. `a` -> `an`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 2]
```diff
@@ -622,6 +622,8 @@ object FunctionRegistry {
     expression[Sha1]("sha", true),
     expression[Sha1]("sha1"),
     expression[Sha2]("sha2"),
+    expression[AesEncrypt]("aes_encrypt"),
+    expression[AesDecrypt]("aes_decrypt"),
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:626]: This seems to be supported by MySQL and Oracle, right? Could you mention it in the PR description?

**Review by dongjoon-hyun — COMMENTED** [review]
BTW, we revert this here before.
- 6624a588c1

Since this has a narrow-downed scope, we may reconsider it.

**Comment by dongjoon-hyun** [comment]
cc @rxin

**Comment by sarutak** [comment]
@dongjoon-hyun Thank you for letting me know. I didn't notice the previous PR. I'll check it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 2 above)
Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:626]: Thanks. I've updated.

**Review by cloud-fan — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you, @sarutak and @cloud-fan .

**Pull request closed without merging**